### DISCUSSION TITLE
Delete image tag from Github Container Registry

### DISCUSSION THREAD

**Discussion by pluma** [community/community#26267] [Packages]
<p>I’m using the new GHCR and generating docker images for pull requests so I can deploy to a staging environment per-PR. This works great using GitHub actions but I can’t find a way to delete the generated image tags when a PR is merged.</p>
<p>Based on the <a href="https://docs.docker.com/registry/spec/api/#delete-manifest" rel="noopener nofollow ugc">Docker Registry API spec</a> it looks like I would need to fetch the manifest digest for the given tag and then delete the manifest. But trying to do this via the GHCR API gives me <code>UNSUPPORTED</code>.</p>
<p>The <a href="https://docs.github.com/en/packages/guides/deleting-a-container-image" rel="noopener nofollow ugc">GHCR doc*mentation</a> only describes the process of manually deleting an image and indicates that it may be added to the GraphQL API in the future.</p>
<p>Is there currently a way to delete image tags via code? If not, is this going to be part of the GA release?</p>
<p>EDIT: I had to censor the word “doc*mentation” because the word filter complained that my post has the word “c*m” in it. I guess someone wasn’t aware of the Sc*nthorpe problem (yeah, that triggers it too).</p>

**Comment by pluma** [comment]
<p>I <em>think</em> the correct solution for the moment is to use the GitHub REST API and delete the <code>container</code> type package with the container name. I haven’t tried this yet, but it works for the older <code>docker</code> type containers (although unintuitively the API doesn’t use the repo name as apparently that’s not actually part of the package name).</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>The API for that was added very recently:</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/" target="_blank" rel="noopener">The GitHub Blog</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/03/blog-card.png?fit=5001%2C2626" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/" target="_blank" rel="noopener">Packages: delete and restore for all packages - GitHub Changelog</a></h3>

<p>Packages: delete and restore for all packages</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I’ve tried the API with a script to clean up old untagged versions, maybe you can get some inspiration from there:</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py" target="_blank" rel="noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py" target="_blank" rel="noopener">airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py</a></h4>
<pre><code class="lang-py">#!/usr/bin/python3
# PYTHON_ARGCOMPLETE_OK
import argparse
import dateutil.parser
import getpass
import os
import requests
from datetime import datetime, timedelta

__author__ = "Fiona Klute"
__version__ = "0.1"
__copyright__ = "Copyright (C) 2021 Fiona Klute"
__license__ = "MIT"

# GitHub API documentation: https://docs.github.com/en/rest/reference/packages
github_api_accept = 'application/vnd.github.v3+json'


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
</code></pre>

  This file has been truncated. <a href="https://github.com/airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py" target="_blank" rel="noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Reply by matifali** [reply]
How can I only delete a specific tag of the image and not the image?

**Reply by rafalkk** [reply]
I also had this problem. You can try my custom action:
https://github.com/rafalkk/remove-dockertag-action

**Comment by pluma** [comment]
<p>That explains why I didn’t find it when I first checked. Looks like it’s still not possible to do this via the Docker (<a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a>) API though. At least with the rest API I can do this alongside the other post-PR cleanup stuff.</p>

**Comment by mr-linch** [comment]
<p>If you need delete images by tag you can checkout our action</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/bots-house/ghcr-delete-image-action" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/3163427/181096556-1ebc5b32-db04-4bc9-98a5-a6b02b00d191.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f_2_10x10.png"></div>

<h3><a href="https://github.com/bots-house/ghcr-delete-image-action" target="_blank" rel="noopener nofollow ugc">bots-house/ghcr-delete-image-action</a></h3>


  <p><span class="label1">♻️ Delete GitHub Container Registry image by tag. Contribute to bots-house/ghcr-delete-image-action development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by saheerb** [comment]
<p>I can delete the image using the APIs, is there anyway to just delete the tag?</p>

**Comment by TonyRL** [comment]
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/marketplace/actions/ghcr-pruning">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/marketplace/actions/ghcr-pruning" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/11386903/181096559-323a688b-cbc2-4e1e-b692-f05f37679f15.jpeg" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/marketplace/actions/ghcr-pruning" target="_blank" rel="noopener nofollow ugc">GHCR Pruning - GitHub Marketplace</a></h3>

  <p>GitHub Action to prune/delete container versions from GitHub Container Registry (ghcr.io)</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>This action supports removing images by tags and also untagged images.</p>

**Comment by Yikun** [comment]
<p>Looks like ghcr didn’t support delete the tag by api?</p>
<p>I also take a look on above actions, it seems the “remove image by tag action” not only cleanup the tag, but also clean the package version. That means if a package version with multi-tags will also be removed.</p>

**Comment by felfert** [comment]
Just stumbled over this::

[This](https://docs.github.com/en/rest/packages?apiVersion=2022-11-28#list-packages-for-the-authenticated-users-namespace) lists the names of containers like ``reponame/tagname`` (Note the slash)
In order to make the [DELETE](https://docs.github.com/en/rest/packages?apiVersion=2022-11-28#delete-a-package-for-a-user--code-samples) work, you have to URL-encode the name. E.g. replace the slash with ``%2f``
Otherwise, you get a 404 not found

**Comment by neilime** [comment]
> Just stumbled over this::
> 
> [This](https://docs.github.com/en/rest/packages?apiVersion=2022-11-28#list-packages-for-the-authenticated-users-namespace) lists the names of containers like `reponame/tagname` (Note the slash) In order to make the [DELETE](https://docs.github.com/en/rest/packages?apiVersion=2022-11-28#delete-a-package-for-a-user--code-samples) work, you have to URL-encode the name. E.g. replace the slash with `%2f` Otherwise, you get a 404 not found

@felfert do you mean you are able to delete tag with this kind of request?

```
DELETE https://api.github.com/user/packages/container/test-repo%2Fapplication/0.5.0
```

I've tried to do same with an organization but have a 404

```
DELETE https://api.github.com/orgs/test-org/packages/container/test-repo%2Fapplication/0.5.0
```

**Reply by felfert** [reply]
No. Not the slash between the repo and the name. The slash **IN** the name (which separes the base name from the tag name)
In your example this would be:

`DELETE https://api.github.com/orgs/test-org/packages/container/test-repo/application%2f0.5.0`

**Reply by neilime** [reply]
It does not work:

Getting package info need to url encode repo name:

- `https://api.github.com/orgs/test-org/packages/container/test-repo%2Fapplication`: 200
   ```json
   {
    "id": 0000000,
    "name": "test-repo/application",
    "package_type": "container",
   }
   ```
Without encoded name issuing a 404:

- `https://api.github.com/orgs/test-org/packages/container/test-repo/application`: 404

Unable to retrieve specific tag for a given package (with and without url encoding):

- `https://api.github.com/orgs/test-org/packages/container/test-repo%2Fapplication/0.5.0` : 404
- `https://api.github.com/orgs/test-org/packages/container/test-repo%2Fapplication%2F0.5.0` : 404
- `https://api.github.com/orgs/test-org/packages/container/test-repo/application%2F0.5.0` : 404

**Comment by rafalkk** [comment]
You can try my custom action:
https://github.com/rafalkk/remove-dockertag-action

**Reply by ryuheechul** [reply]
> You can try my custom action: https://github.com/rafalkk/remove-dockertag-action

Same idea as the action above, but a workaround using Podman (or Docker that is authenticated to ghcr.io so a write access already configured).

```bash
# these are optional but I recommend
set -e # so it stops on unexpected error
set -x # so you know what commands runs and maps the stdout/stderr to them

# 0. Change this to match the tag to delete
tag_to_delete="ghcr.io/OWNER/REPO:TAG_TO_DELETE"

# 1. Create a unique string file
echo "random-unlink-token-$(date +%s%N)" > unique-trigger.txt && cat unique-trigger.txt

# 2-0. Optional: clean up any pre-existing one just in case
podman artifact rm -i "${tag_to_delete}"
# 2. Create a image based on the generated file
podman artifact add "${tag_to_delete}" unique-trigger.txt
# 2-1. Verify the tag is set to a small image
podman image list "${tag_to_delete}"

# 3. Push it to GHCR trigger unlinking the tag from previous manifest
podman push "${tag_to_delete}"
```

So running above should move the tag from the original manifest to the newly pushed unique one.

And now you can visit `https://github.com/OWNER/REPO/pkgs/container/PACKAGENAME/versions`. 

> [!WARNING]
> Make sure this tag that you are trying to delete is solely attached to the manifest that is safe to delete first.

And delete the version that is attached to the tag (which will delete the manifest attach to the tag not just the tags - which is the whole reason why this discussion thread exist).


While this is a quick and dirty manual approach, it's straight forward and doesn't require any other setup (assuming podman/docker is already configured) which is suitable for a one-off operation.
